### PR TITLE
Dynamic output fields

### PR DESCRIPTION
**Pull request by mankenavenkatesh**
## PR Description
**Changes and workflow**
1. Output Fields for events are added in config.toml file under [output.fields] section with a getter inside ArtemisConfiguration. 
2. These fields are added to a lookup map called "eventOutputFields" inside EventHandler.class by fetching from ArtemisConfiguration.
3. When event is triggered, this map "eventOutputFields" is looked up for fields to output and "filterOutputFields" method is called on triggered event Object.
4. This method will compute the fields and it's values to output and stores it in the state called "outputFieldMap" inside event Object.
5. Based on providerType[JSON, CSV], respective methods toJson() or toCSV() is called. These methods will refer to  "outputFieldMap" to generate the respective output Type.

### CODE REVIEW

File: data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java [hunk 1]
```diff
@@ -19,10 +19,15 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.SerializerProvider;
 import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonObject;
 import java.io.IOException;
```
Comment (jrhea) [inline data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java:22]: was it not possible to implement using jackson?  It's kind of weird to have to ways to implement json here

File: pow/src/main/java/tech/pegasys/artemis/pow/event/Deposit.java [hunk 2]
```diff
@@ -78,37 +82,56 @@ public long getAmount() {
...
 
   @Override
   public String toCSV() {
-    return null;
+    String csvOutputString = "";
+    for (Object obj : this.outputFieldMap.values()) {
+      csvOutputString += "'" + obj.toString() + "',";
+    }
+    return csvOutputString.substring(0, csvOutputString.length() - 1);
   }
```
Comment (jrhea) [inline pow/src/main/java/tech/pegasys/artemis/pow/event/Deposit.java:132]: Need a check to ensure that csvOutputString.length() > 0

File: pow/src/main/java/tech/pegasys/artemis/pow/event/Eth2Genesis.java [hunk 3]
```diff
@@ -49,28 +53,47 @@ public long getTime() {
...
   }
 
   @Override
   public String toCSV() {
-    return null;
+    String csvOutputString = "";
+    for (Object obj : this.outputFieldMap.values()) {
+      csvOutputString += "'" + obj.toString() + "',";
+    }
+    return csvOutputString.substring(0, csvOutputString.length() - 1);
```
Comment (jrhea) [inline pow/src/main/java/tech/pegasys/artemis/pow/event/Eth2Genesis.java]: Need a check to ensure that csvOutputString.length() > 0

File: validator/client/src/main/java/tech/pegasys/artemis/validator/client/DepositSimulation.java [hunk 4]
```diff
@@ -76,19 +70,56 @@ public String toJSON() {
...
 
   @Override
   public String toCSV() {
-    return null;
+    String csvOutputString = "";
+    for (Object obj : this.outputFieldMap.values()) {
+      csvOutputString += "'" + obj.toString() + "',";
+    }
+    return csvOutputString.substring(0, csvOutputString.length() - 1);
   }
```
Comment (jrhea) [inline validator/client/src/main/java/tech/pegasys/artemis/validator/client/DepositSimulation.java:130]: Need a check to ensure that csvOutputString.length() > 0

File: data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java [hunk 5]
```diff
@@ -114,62 +120,100 @@ public TimeSeriesRecord(
...
   }
 
+  @Override
+  public String toJSON() {
+    Gson gson = new GsonBuilder().create();
+    GsonBuilder gsonBuilder = new GsonBuilder();
+    String jsonString = gson.toJson(this.outputFieldMap);
+    JsonObject timeSeriesRecord = gson.fromJson(jsonString, JsonObject.class);
+    Gson customGson = gsonBuilder.setPrettyPrinting().create();
+    return customGson.toJson(timeSeriesRecord);
```
Comment (jrhea) [inline data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java]: i think pretty printing is supposed to be optional.  @dangerousfood isn't that what the `format` option is for?

**Review by jrhea — CHANGES_REQUESTED** [review]
Thanks for the PR!!  I added some comments.  Also, two tests are failing:

```
tech.pegasys.artemis.data.provider.CSVProviderTest > testCSVPrint(Path) FAILED
    java.lang.StringIndexOutOfBoundsException at CSVProviderTest.java:37

tech.pegasys.artemis.data.provider.JSONProviderTest > testJSONPrint(Path) FAILED
    java.lang.NullPointerException at JSONProviderTest.java:45
```

those tests should be operational before merging.  Thanks!!!

File: data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java (hunk 5 above)
Comment (dangerousfood) [reply data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java]: Nah, pretty printing just brings in tabs and new line characters. You're thinking of the formatted output that puts brackets around the whole thing

File: data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java (hunk 1 above)
Comment (mankenavenkatesh) [reply data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java:22]: Yeah...that's much simpler as well. 
Modified code with required changes. Please review.

File: pow/src/main/java/tech/pegasys/artemis/pow/event/Deposit.java (hunk 2 above)
Comment (mankenavenkatesh) [reply pow/src/main/java/tech/pegasys/artemis/pow/event/Deposit.java:132]: My Mistake. Fixed it.

File: pow/src/main/java/tech/pegasys/artemis/pow/event/Eth2Genesis.java (hunk 3 above)
Comment (mankenavenkatesh) [reply pow/src/main/java/tech/pegasys/artemis/pow/event/Eth2Genesis.java]: Done. please review.

File: validator/client/src/main/java/tech/pegasys/artemis/validator/client/DepositSimulation.java (hunk 4 above)
Comment (mankenavenkatesh) [reply validator/client/src/main/java/tech/pegasys/artemis/validator/client/DepositSimulation.java:130]: Done. please review.

File: pow/src/main/java/tech/pegasys/artemis/pow/event/Deposit.java (hunk 2 above)
Comment (jrhea) [reply pow/src/main/java/tech/pegasys/artemis/pow/event/Deposit.java:132]: 👍

File: data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java (hunk 1 above)
Comment (jrhea) [reply data/src/main/java/tech/pegasys/artemis/data/TimeSeriesRecord.java:22]: 🎉 Excellent! Thanks!!

File: pow/src/main/java/tech/pegasys/artemis/pow/event/Eth2Genesis.java (hunk 3 above)
Comment (jrhea) [reply pow/src/main/java/tech/pegasys/artemis/pow/event/Eth2Genesis.java]: 👍

File: validator/client/src/main/java/tech/pegasys/artemis/validator/client/DepositSimulation.java (hunk 4 above)
Comment (jrhea) [reply validator/client/src/main/java/tech/pegasys/artemis/validator/client/DepositSimulation.java:130]: 👍

**Review by jrhea — APPROVED** [review]
Great work!  Thank you for doing this!!!!!!

**Pull request merged**